File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章四平方和.cpp
// 四平方和  2016年第七届省赛  lanqiaoOJ 122
#include<bits/stdc++.h>
using namespace std;
int main(){
    int n; cin>>n;
    for(int a=0;a<=sqrt(n);a++)                //限制a的范围
        for(int b=a;a*a+b*b<=n;b++)            //限制b的范围
            for(int c=b;a*a+b*b+c*c<=n;c++){   //限制c的范围
                int t = n-a*a-b*b-c*c;
                int d = sqrt(t);               //计算出d
                if(d*d==t && d>=c){
                   cout<<a<<" "<<b<<" "<<c<<" "<<d<<"\n";
                   return 0;
                }
            }
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章四平方和.py
#四平方和  2016年第七届省赛  lanqiaoOJ 122

from math import *
try:
    n = int(input())
    n_1 = int(sqrt(n))
    for a in range(n_1+1):
        n_2 = int(sqrt(n-a*a))
        for b in range(a,n_2+1):
            n_3 = int(sqrt(n-a*a-b*b))
            for c in range(b,n_3+1):
                t = n-a*a-b*b-c*c
                d = int(sqrt(t))
                if d<c:  break
                if d*d==t: print(a,b,c,d); exit(0)         
except Exception as e: pass   #如果sqrt(i)的i是负数，忽略
